File: backend/services/ai_service.py
"""
AI service for document processing, question extraction, and response generation.
"""
import asyncio
import json
from typing import List, Dict, Any, Optional
from uuid import UUID, uuid4
from datetime import datetime

from models import (
    ExtractQuestionsRequest, ExtractQuestionsResponse, Section, Question,
    GenerateResponseRequest, GenerateResponseResponse,
    MultiStepResponse, StepResult, StepType, StepStatus,
    QuestionAnalysis, QuestionComplexity, DocumentSearchResult
)

class AIService:
    """Service for AI-powered document processing and response generation."""
    
    def __init__(self):
        # Configuration
        self.openai_api_key = None  # Set from environment
        self.llamacloud_api_key = None  # Set from environment
        
    def configure(self, openai_api_key: str = None, llamacloud_api_key: str = None):
        """Configure AI service with API keys."""
        if openai_api_key:
            self.openai_api_key = openai_api_key
        if llamacloud_api_key:
            self.llamacloud_api_key = llamacloud_api_key
    
    async def extract_questions(self, request: ExtractQuestionsRequest) -> ExtractQuestionsResponse:
        """
        Extract questions from RFP document content using AI.
        This simulates the AI processing - in production, would use OpenAI API.
        """
        # Simulate AI processing delay
        await asyncio.sleep(1)
        
        # Mock extracted questions - in production, would use OpenAI to parse document
        sections = []
        
        # Example extracted sections based on common RFP structure
        if "technical" in request.content.lower() or "requirements" in request.content.lower():
            technical_section = Section(
                id="section_technical",
                title="Technical Requirements",
                description="Questions related to technical capabilities and requirements",
                questions=[
                    Question(
                        id=uuid4(),
                        reference_id="question_1.1",
                        text="Describe your company's technical infrastructure and capabilities.",
                        topic="Technical Infrastructure",
                        project_id=request.project_id
                    ),
                    Question(
                        id=uuid4(),
                        reference_id="question_1.2", 
                        text="What security measures and protocols does your organization implement?",
                        topic="Security",
                        project_id=request.project_id
                    ),
                    Question(
                        id=uuid4(),
                        reference_id="question_1.3",
                        text="Describe your data backup and disaster recovery procedures.",
                        topic="Data Management",
                        project_id=request.project_id
                    )
                ]
            )
            sections.append(technical_section)
        
        if "experience" in request.content.lower() or "qualifications" in request.content.lower():
            experience_section = Section(
                id="section_experience",
                title="Company Experience & Qualifications",
                description="Questions about company background and experience",
                questions=[
                    Question(
                        id=uuid4(),
                        reference_id="question_2.1",
                        text="Provide details about your company's relevant experience in this domain.",
                        topic="Experience",
                        project_id=request.project_id
                    ),
                    Question(
                        id=uuid4(),
                        reference_id="question_2.2",
                        text="List key personnel who will be involved in this project and their qualifications.",
                        topic="Personnel",
                        project_id=request.project_id
                    ),
                    Question(
                        id=uuid4(),
                        reference_id="question_2.3",
                        text="Provide references from similar projects completed in the last 3 years.",
                        topic="References",
                        project_id=request.project_id
                    )
                ]
            )
            sections.append(experience_section)
        
        if "pricing" in request.content.lower() or "cost" in request.content.lower():
            pricing_section = Section(
                id="section_pricing",
                title="Pricing & Commercial Terms",
                description="Questions related to pricing and commercial aspects",
                questions=[
                    Question(
                        id=uuid4(),
                        reference_id="question_3.1",
                        text="Provide detailed pricing breakdown for all proposed services.",
                        topic="Pricing",
                        project_id=request.project_id
                    ),
                    Question(
                        id=uuid4(),
                        reference_id="question_3.2",
                        text="What are your payment terms and conditions?",
                        topic="Payment Terms",
                        project_id=request.project_id
                    )
                ]
            )
            sections.append(pricing_section)
        
        # If no specific content detected, create a general section
        if not sections:
            general_section = Section(
                id="section_general",
                title="General Requirements",
                description="General questions extracted from the document",
                questions=[
                    Question(
                        id=uuid4(),
                        reference_id="question_1.1",
                        text="Describe your company's approach to this project.",
                        topic="General Approach",
                        project_id=request.project_id
                    ),
                    Question(
                        id=uuid4(),
                        reference_id="question_1.2",
                        text="What makes your company uniquely qualified for this opportunity?",
                        topic="Qualifications",
                        project_id=request.project_id
                    )
                ]
            )
            sections.append(general_section)
        
        return ExtractQuestionsResponse(
            document_id=request.document_id,
            document_name=request.document_name,
            sections=sections,
            extracted_at=datetime.now()
        )
    
    async def generate_response(self, request: GenerateResponseRequest) -> GenerateResponseResponse:
        """
        Generate AI response for a question.
        This simulates basic response generation - in production, would use OpenAI API.
        """
        # Simulate AI processing delay
        await asyncio.sleep(2)
        
        # Mock response generation
        question_lower = request.question.lower()
        
        # Generate contextual response based on question content
        if "technical" in question_lower or "infrastructure" in question_lower:
            response = """Our company maintains state-of-the-art technical infrastructure including:

• Cloud-native architecture built on Microsoft Azure with 99.9% uptime SLA
• Microservices-based application design for scalability and maintainability  
• Automated CI/CD pipelines using Azure DevOps for rapid deployment
• Container orchestration with Azure Kubernetes Service (AKS)
• Comprehensive monitoring and logging with Azure Monitor and Application Insights

Our infrastructure is designed to handle high-volume transactions while maintaining security and performance standards."""
            
        elif "security" in question_lower:
            response = """Our organization implements comprehensive security measures:

• ISO 27001 and SOC 2 Type II certified security management
• Multi-factor authentication (MFA) for all system access
• End-to-end encryption for data in transit and at rest
• Regular penetration testing and vulnerability assessments
• 24/7 security operations center (SOC) monitoring
• Compliance with GDPR, HIPAA, and industry-specific regulations

We follow zero-trust security principles and maintain detailed audit logs for all system activities."""
            
        elif "experience" in question_lower:
            response = """Our company brings extensive relevant experience:

• 15+ years in the industry with over 200 successful project implementations
• Dedicated team of certified professionals with domain expertise
• Proven track record with Fortune 500 companies and government agencies
• Average project success rate of 98% with on-time, on-budget delivery
• Strong partnerships with leading technology providers
• Continuous investment in training and certification programs

We have successfully delivered similar projects across various industries including healthcare, finance, and manufacturing."""
            
        elif "pricing" in question_lower or "cost" in question_lower:
            response = """Our pricing structure is transparent and competitive:

• Fixed-price model for defined scope with no hidden costs
• Detailed breakdown of all components and deliverables
• Volume discounts available for multi-year commitments
• Flexible payment terms aligned with project milestones
• ROI guarantee with measurable performance metrics
• Optional support and maintenance packages available

We provide detailed cost justification and total cost of ownership analysis."""
            
        else:
            response = f"""Thank you for your question: "{request.question}"

We understand the importance of this requirement and are committed to providing a comprehensive solution. Our approach includes:

• Thorough analysis of your specific needs and requirements
• Custom solution design tailored to your organization
• Experienced team dedicated to project success
• Proven methodologies and best practices
• Ongoing support and optimization
• Transparent communication throughout the project lifecycle

We would welcome the opportunity to discuss this in more detail and provide additional information as needed."""
        
        # Mock sources
        sources = [
            {
                "id": 1,
                "fileName": "company_capabilities.pdf",
                "pageNumber": "3-5",
                "relevance": 95,
                "textContent": "Relevant excerpt from company capabilities document..."
            },
            {
                "id": 2,
                "fileName": "technical_architecture.docx", 
                "pageNumber": "12",
                "relevance": 88,
                "textContent": "Technical architecture overview and specifications..."
            }
        ]
        
        metadata = {
            "confidence": 0.92,
            "generatedAt": datetime.now().isoformat(),
            "indexesUsed": request.index_ids,
            "note": "Response generated using company knowledge base"
        }
        
        return GenerateResponseResponse(
            success=True,
            response=response,
            sources=sources,
            metadata=metadata
        )
    
    async def multi_step_generate_response(self, request: GenerateResponseRequest) -> MultiStepResponse:
        """
        Generate response using multi-step AI analysis process.
        """
        response_id = str(uuid4())
        steps = []
        start_time = datetime.now()
        
        # Step 1: Analyze Question
        step1 = await self._analyze_question_step(request.question)
        steps.append(step1)
        
        # Step 2: Search Documents
        step2 = await self._search_documents_step(request.question, request.index_ids)
        steps.append(step2)
        
        # Step 3: Extract Information
        step3 = await self._extract_information_step(request.question, step2.output)
        steps.append(step3)
        
        # Step 4: Synthesize Response
        step4 = await self._synthesize_response_step(request.question, step3.output)
        steps.append(step4)
        
        # Step 5: Validate Answer
        step5 = await self._validate_answer_step(step4.output["response"])
        steps.append(step5)
        
        end_time = datetime.now()
        total_duration = (end_time - start_time).total_seconds()
        
        return MultiStepResponse(
            id=response_id,
            question_id=request.question_id,
            project_id=request.project_id,
            steps=steps,
            final_response=step4.output["response"],
            overall_confidence=step5.output["overall_confidence"],
            total_duration=total_duration,
            sources=step4.output["sources"],
            metadata={
                "steps_completed": len(steps),
                "processing_start_time": start_time,
                "processing_end_time": end_time,
                "model_used": "gpt-4o",
                "tokens_used": 2500
            }
        )
    
    async def _analyze_question_step(self, question: str) -> StepResult:
        """Step 1: Analyze the question complexity and requirements."""
        step_start = datetime.now()
        await asyncio.sleep(0.5)  # Simulate processing
        
        # Mock analysis
        complexity = QuestionComplexity.MODERATE
        if len(question.split()) > 20:
            complexity = QuestionComplexity.COMPLEX
        elif "?" in question and len(question.split("?")) > 2:
            complexity = QuestionComplexity.MULTI_PART
        
        analysis = QuestionAnalysis(
            complexity=complexity,
            required_information=["company capabilities", "technical specifications", "experience"],
            specific_entities=["infrastructure", "security", "team"],
            search_queries=["technical capabilities", "security measures", "project experience"],
            expected_sources=3,
            reasoning="Question requires comprehensive response covering multiple technical aspects"
        )
        
        step_end = datetime.now()
        duration = (step_end - step_start).total_seconds()
        
        return StepResult(
            id="step_analyze_question",
            type=StepType.ANALYZE_QUESTION,
            title="Analyze Question",
            description="Analyzing question complexity and requirements",
            status=StepStatus.COMPLETED,
            start_time=step_start,
            end_time=step_end,
            duration=duration,
            output=analysis.model_dump(),
            metadata={"complexity_score": 0.7}
        )
    
    async def _search_documents_step(self, question: str, index_ids: List[str]) -> StepResult:
        """Step 2: Search relevant documents."""
        step_start = datetime.now()
        await asyncio.sleep(1.0)  # Simulate processing
        
        # Mock search results
        search_result = DocumentSearchResult(
            query=question,
            documents_found=5,
            relevant_sources=[
                {
                    "id": "doc1", 
                    "title": "Technical Capabilities Overview",
                    "relevanceScore": 0.95,
                    "snippet": "Our technical infrastructure includes..."
                },
                {
                    "id": "doc2",
                    "title": "Security Protocols and Procedures", 
                    "relevanceScore": 0.89,
                    "snippet": "Security measures implemented include..."
                }
            ],
            coverage="complete"
        )
        
        step_end = datetime.now()
        duration = (step_end - step_start).total_seconds()
        
        return StepResult(
            id="step_search_documents",
            type=StepType.SEARCH_DOCUMENTS,
            title="Search Documents",
            description="Searching relevant documents in knowledge base",
            status=StepStatus.COMPLETED,
            start_time=step_start,
            end_time=step_end,
            duration=duration,
            output=search_result.model_dump(),
            metadata={"indexes_searched": index_ids}
        )
    
    async def _extract_information_step(self, question: str, search_output: Dict[str, Any]) -> StepResult:
        """Step 3: Extract relevant information."""
        step_start = datetime.now()
        await asyncio.sleep(0.8)  # Simulate processing
        
        # Mock information extraction
        extraction_result = {
            "extracted_facts": [
                {
                    "fact": "Company has 99.9% uptime SLA",
                    "source": "Technical Capabilities Overview",
                    "confidence": 0.95
                },
                {
                    "fact": "ISO 27001 certified security management",
                    "source": "Security Protocols and Procedures",
                    "confidence": 0.92
                }
            ],
            "missing_information": [],
            "conflicting_information": []
        }
        
        step_end = datetime.now()
        duration = (step_end - step_start).total_seconds()
        
        return StepResult(
            id="step_extract_information",
            type=StepType.EXTRACT_INFORMATION,
            title="Extract Information",
            description="Extracting relevant facts and information",
            status=StepStatus.COMPLETED,
            start_time=step_start,
            end_time=step_end,
            duration=duration,
            output=extraction_result,
            metadata={"facts_extracted": len(extraction_result["extracted_facts"])}
        )
    
    async def _synthesize_response_step(self, question: str, extraction_output: Dict[str, Any]) -> StepResult:
        """Step 4: Synthesize final response."""
        step_start = datetime.now()
        await asyncio.sleep(1.2)  # Simulate processing
        
        # Generate comprehensive response
        response = await self.generate_response(GenerateResponseRequest(
            question=question,
            question_id=uuid4(),
            project_id=uuid4(),
            index_ids=[]
        ))
        
        synthesis_result = {
            "response": response.response,
            "confidence": 0.91,
            "sources": response.sources,
            "recommendations": [
                "Consider including specific metrics and benchmarks",
                "Provide detailed implementation timeline"
            ]
        }
        
        step_end = datetime.now()
        duration = (step_end - step_start).total_seconds()
        
        return StepResult(
            id="step_synthesize_response",
            type=StepType.SYNTHESIZE_RESPONSE,
            title="Synthesize Response",
            description="Synthesizing comprehensive response",
            status=StepStatus.COMPLETED,
            start_time=step_start,
            end_time=step_end,
            duration=duration,
            output=synthesis_result,
            metadata={"response_length": len(synthesis_result["response"])}
        )
    
    async def _validate_answer_step(self, response: str) -> StepResult:
        """Step 5: Validate the generated answer."""
        step_start = datetime.now()
        await asyncio.sleep(0.3)  # Simulate processing
        
        # Mock validation
        validation_result = {
            "overall_confidence": 0.91,
            "completeness_score": 0.89,
            "accuracy_score": 0.93,
            "relevance_score": 0.92,
            "validation_checks": [
                {"check": "Response addresses question directly", "passed": True},
                {"check": "Sources are relevant and credible", "passed": True},
                {"check": "Information is factually consistent", "passed": True},
                {"check": "Response is comprehensive", "passed": True}
            ],
            "recommendations": [
                "Response meets quality standards",
                "Ready for delivery"
            ]
        }
        
        step_end = datetime.now()
        duration = (step_end - step_start).total_seconds()
        
        return StepResult(
            id="step_validate_answer",
            type=StepType.VALIDATE_ANSWER,
            title="Validate Answer",
            description="Validating response quality and accuracy",
            status=StepStatus.COMPLETED,
            start_time=step_start,
            end_time=step_end,
            duration=duration,
            output=validation_result,
            metadata={"validation_passed": True}
        )


# Global service instance
ai_service = AIService()

File: frontend/src/app/services/ai.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, BehaviorSubject } from 'rxjs';
import {
  ExtractQuestionsRequest,
  ExtractQuestionsResponse,
  GenerateResponseRequest,
  GenerateResponseResponse,
  MultiStepResponse,
} from '../models';

@Injectable({
  providedIn: 'root',
})
export class AIService {
  private baseUrl = 'http://localhost:8000';
  private processingStepsSubject = new BehaviorSubject<any[]>([]);
  public processingSteps$ = this.processingStepsSubject.asObservable();

  constructor(private http: HttpClient) {}

  // Document processing
  extractQuestionsFromDocument(
    request: ExtractQuestionsRequest
  ): Observable<ExtractQuestionsResponse> {
    return this.http.post<ExtractQuestionsResponse>(
      `${this.baseUrl}/ai/extract-questions`,
      request
    );
  }

  // Response generation
  generateResponse(
    request: GenerateResponseRequest
  ): Observable<GenerateResponseResponse> {
    return this.http.post<GenerateResponseResponse>(
      `${this.baseUrl}/ai/generate-response`,
      request
    );
  }

  // Multi-step response generation
  generateMultiStepResponse(
    request: GenerateResponseRequest
  ): Observable<MultiStepResponse> {
    return this.http.post<MultiStepResponse>(
      `${this.baseUrl}/ai/multi-step-response`,
      request
    );
  }

  // Processing state management
  setProcessingSteps(steps: any[]): void {
    this.processingStepsSubject.next(steps);
  }

  getProcessingSteps(): any[] {
    return this.processingStepsSubject.value;
  }

  clearProcessingSteps(): void {
    this.processingStepsSubject.next([]);
  }

  // Utility methods for AI processing
  getStepIcon(stepType: string): string {
    switch (stepType) {
      case 'analyze_question':
        return 'psychology';
      case 'search_documents':
        return 'search';
      case 'extract_information':
        return 'fact_check';
      case 'synthesize_response':
        return 'auto_awesome';
      case 'validate_answer':
        return 'check_circle';
      default:
        return 'settings';
    }
  }

  getStepColor(status: string): string {
    switch (status) {
      case 'pending':
        return 'warn';
      case 'running':
        return 'accent';
      case 'completed':
        return 'primary';
      case 'failed':
        return 'warn';
      default:
        return 'basic';
    }
  }

  formatStepDuration(duration?: number): string {
    if (!duration) return '';
    return `${duration.toFixed(2)}s`;
  }

  calculateOverallProgress(steps: any[]): number {
    if (steps.length === 0) return 0;

    const completedSteps = steps.filter(
      (step) => step.status === 'completed'
    ).length;
    return (completedSteps / steps.length) * 100;
  }

  getConfidenceColor(confidence: number): string {
    if (confidence >= 0.8) return 'primary';
    if (confidence >= 0.6) return 'accent';
    return 'warn';
  }

  getConfidenceText(confidence: number): string {
    if (confidence >= 0.9) return 'Very High';
    if (confidence >= 0.8) return 'High';
    if (confidence >= 0.6) return 'Medium';
    if (confidence >= 0.4) return 'Low';
    return 'Very Low';
  }

  // Mock data generators for development/demo
  generateMockQuestionExtractionResult(
    documentName: string,
    projectId: string
  ): ExtractQuestionsResponse {
    const mockSections = [
      {
        id: 'section_technical',
        title: 'Technical Requirements',
        description: 'Questions related to technical capabilities',
        questions: [
          {
            id: '1',
            reference_id: 'Q1.1',
            text: 'Describe your technical infrastructure and security measures.',
            topic: 'Infrastructure',
            project_id: projectId,
            created_at: new Date(),
            updated_at: new Date(),
          },
          {
            id: '2',
            reference_id: 'Q1.2',
            text: 'What backup and disaster recovery procedures do you have in place?',
            topic: 'Business Continuity',
            project_id: projectId,
            created_at: new Date(),
            updated_at: new Date(),
          },
        ],
      },
      {
        id: 'section_experience',
        title: 'Company Experience',
        description: 'Questions about company background and qualifications',
        questions: [
          {
            id: '3',
            reference_id: 'Q2.1',
            text: 'Provide details about your relevant experience in this domain.',
            topic: 'Experience',
            project_id: projectId,
            created_at: new Date(),
            updated_at: new Date(),
          },
        ],
      },
    ];

    return {
      document_id: 'mock-doc-id',
      document_name: documentName,
      sections: mockSections,
      extracted_at: new Date(),
    };
  }

  generateMockResponseResult(question: string): GenerateResponseResponse {
    const mockResponse = `Based on your question: "${question}"

Our company provides comprehensive solutions that address your requirements:

• **Technical Excellence**: State-of-the-art infrastructure with 99.9% uptime SLA
• **Security First**: ISO 27001 certified with multi-layered security protocols  
• **Proven Experience**: 15+ years serving clients across various industries
• **Expert Team**: Certified professionals with deep domain expertise
• **Quality Assurance**: Rigorous testing and validation processes

We are committed to delivering exceptional results that exceed your expectations.`;

    return {
      success: true,
      response: mockResponse,
      sources: [
        {
          id: 1,
          fileName: 'company_capabilities.pdf',
          pageNumber: '3-5',
          relevance: 95,
          textContent: 'Technical infrastructure overview...',
        },
        {
          id: 2,
          fileName: 'security_protocols.docx',
          pageNumber: '12',
          relevance: 88,
          textContent: 'Security measures and compliance...',
        },
      ],
      metadata: {
        confidence: 0.92,
        generatedAt: new Date().toISOString(),
        indexesUsed: ['index-1', 'index-2'],
        note: 'Generated using company knowledge base',
      },
    };
  }

  // File processing utilities
  validateFileForProcessing(file: File): { valid: boolean; error?: string } {
    const maxSize = 50 * 1024 * 1024; // 50MB
    const allowedTypes = [
      'application/pdf',
      'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
      'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
      'application/vnd.openxmlformats-officedocument.presentationml.presentation',
      'text/plain',
    ];

    if (file.size > maxSize) {
      return { valid: false, error: 'File size exceeds 50MB limit' };
    }

    if (!allowedTypes.includes(file.type)) {
      return {
        valid: false,
        error:
          'Unsupported file type. Please upload PDF, Word, Excel, PowerPoint, or text files.',
      };
    }

    return { valid: true };
  }

  async readFileContent(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.onload = (e) => {
        resolve(e.target?.result as string);
      };
      reader.onerror = (e) => {
        reject(new Error('Failed to read file'));
      };
      reader.readAsText(file);
    });
  }
}
